pybamm/solvers/idaklu_jax.py

```diff
@@ -0,0 +1,956 @@
+        y_dot = np.array(y_dot)
+        return y_dot
+
+    def _jax_vjp_impl_array_inputs(
+        self,
+        y_bar,
+        y_bar_s0,
+        y_bar_s1,
+        invar,
+        primal_t,
+        primal_inputs,
+    ):
+        """Wrapper for VJP implementation used by IDAKLU callback
+
+        This version assumes all parameters are provided as np.ndarray vectors
+        """
+        # Reshape y_bar
+        logging.debug(f"Reshaping y_bar to ({y_bar_s0}, {y_bar_s1})")
+        y_bar = y_bar.reshape(y_bar_s0, y_bar_s1)
+        logging.debug(f"y_bar is now: {y_bar}")
+        primals = primal_t, *tuple([k for k in primal_inputs])
+        return self._jax_vjp_impl(y_bar, invar, *primals)
+
+    def _register_callbacks(self):
+        """Register the solve method with the IDAKLU solver"""
+        logging.info("_register_callbacks")
+        self.idaklu_jax_obj.register_callbacks(
+            self._jax_solve_array_inputs,
+            self._jax_jvp_impl_array_inputs,
+            self._jax_vjp_impl_array_inputs,
+        )
+
+    def _unique_name(self):
+        """Return a unique name for this solver object for naming the JAX primitives"""
+        return f"{self.idaklu_jax_obj.get_index()}"
+
+    def jaxify(
+        self,
+        model,
+        t_eval,
+        *,
+        output_variables=None,
+        inputs=None,
+        calculate_sensitivities=True,
+    ):
+        """JAXify the model and solver
+
+        Creates a JAX expression representing the IDAKLU-wrapped solver
+        object.
+
+        Parameters
+        ----------
+        model : :class:`pybamm.BaseModel`
+            The model to be solved
+        t_eval : numeric type, optional
+            The times at which to compute the solution. If None, the times in the model
+            are used.
+        output_variables : list of str, optional
+            The variables to be returned. If None, the variables in the model are used.
+        inputs : dict, optional
+            Any inputs to the model
+        calculate_sensitivities : bool, optional
+            Whether to calculate sensitivities. Default is True.
+        """
+        if self.jaxpr is not None:
+            warnings.warn(
+                "JAX expression has already been created. "
+                "Overwriting with new expression.",
+                UserWarning,
+            )
+        self.jaxpr = self._jaxify(
+            model,
+            t_eval,
+            output_variables=output_variables,
+            inputs=inputs,
+            calculate_sensitivities=calculate_sensitivities,
+        )
+        return self.jaxpr
+
+    def _jaxify(
+        self,
+        model,
+        t_eval,
+        *,
+        output_variables=None,
+        inputs=None,
+        calculate_sensitivities=True,
+    ):
+        """JAXify the model and solver"""
+
+        self.jax_model = model
+        self.jax_t_eval = t_eval
+        self.jax_output_variables = (
```

Comment:
some of these attributes are essential for most of the functions above, would it work doing the jaxification in the `__init__` so you are sure these attributes are set?

Reply:
Yes, okay. I've jaxified on initialisation.